MudTooltip: Activation via Hover, Focus or Click

mckaragoz opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
Fixes #4564.
First of all, we needed to wait popover fix to see true effects on server-side mobile. It's done, so we can discuss it now.

### So why we need this?
With #4086 Focus event added to the tooltip, but that may be a problem for some situations, because there is no other option to prevent it.

Here is an example. The "GitHub Repository" tooltip doesn't close when you open a new tab and turn the page, because it still has focus and for current situation no way to prevent this.


https://user-images.githubusercontent.com/78308169/169671640-9c24f2bf-ccff-4ed7-b6b3-ec3390183ec7.mp4

Examples can be multiplied. So instead of force users to certain scenarios, adding options could make sense. The other argument, we still can not guarantee to certainly working tooltip/popover for all server-side/wasm projects. So this PR raises the probability of creating workarounds if we have side effects or something other with .net 7 in future.

So its not a breaking change. Tooltip still shows on focus and hover, but now we have options that `ShowOnHover` and `ShowOnFocus`. For the upper example, users can set `ShowOnFocus="false"` to disable the tooltip.

Or, users still want the tooltip with hover. This time they can use `ClickBehavior="TooltipClickBehavior.ClickToHide"`. With this, tooltip shows on focus, but when user click on it, it disappears. It will solve the problems like upper example.

This PR doesn't have a touch addition. I think touch is a more general thing and should be discussed to adding whole library. But users can play with parameters to have their own touch reaction.

The result of this PR:


https://user-images.githubusercontent.com/78308169/169671882-c7c0b390-d87f-466e-a46d-8cf296eaa242.mp4


https://user-images.githubusercontent.com/78308169/169671888-bdb3541c-9674-4700-a4ed-4fb20bcb0b5b.mp4



## How Has This Been Tested?
Tested visually with Server-Side and WASM and the touch versions.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) (fixes bugs, but actually they are not bugs)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

mckaragoz committed:
Tooltip ReactWith

mckaragoz committed:
Click adding

mckaragoz committed:
Text fix

mckaragoz committed:
Some try

mckaragoz committed:
Merge branch 'dev' into TooltipReactWith

mckaragoz committed:
Some interesting code

mckaragoz committed:
Merge branch 'MudBlazor:dev' into TooltipReactWith

mckaragoz committed:
Some More Changes

mckaragoz committed:
Last Changes

mckaragoz committed:
Merge branch 'dev' into TooltipReactWith

henon commented on src/MudBlazor.Docs/Pages/Components/Tooltip/TooltipPage.razor:
```diff
@@ -49,5 +49,20 @@
         </DocsPageSection>
 
+        <DocsPageSection>
+            <SectionHeader Title="Playground For Event Reactions">
```
Rename the title to "Activation Events"

henon commented on src/MudBlazor.Docs/Pages/Components/Tooltip/TooltipPage.razor:
```diff
@@ -49,5 +49,20 @@
 
+        <DocsPageSection>
+            <SectionHeader Title="Playground For Event Reactions">
+                <Description>
+                    With <CodeInline>ShowOnHover</CodeInline> <CodeInline>ShowOnFocus</CodeInline> and <CodeInline>ClickBehavior</CodeInline> parameters, it is possible to choose which events the tooltip will react to.
+                    <br />
+                    Note: if you set <CodeInline>ClickBehavior="TooltipClickBehavior.ClickToShow"</CodeInline>, it overrides and disables hover event reactions.
+                    <br />
+                    Note 2: if you set <CodeInline>ClickBehavior="TooltipClickBehavior.ClickToShow"</CodeInline> and set false to <CodeInline>ShowOnFocus</CodeInline>, tooltip only closes with another click on the item. (Can be useful for some info buttons.)
+                </Description>
```
Change to this:

```
With the <CodeInline>ShowOnHover</CodeInline>, <CodeInline>ShowOnFocus</CodeInline> and <CodeInline>ClickBehavior</CodeInline> parameters, it is possible to choose which events activate and deactivate the tooltip.
                    <br />
                    Note: if you set <CodeInline>ClickBehavior="TooltipClickBehavior.ClickToShow"</CodeInline>, it overrides and disables activation on hover.
                    <br />
                    Note 2: if you set <CodeInline>ClickBehavior="TooltipClickBehavior.ClickToShow"</CodeInline> and set <CodeInline>ShowOnFocus</CodeInline> to false, the tooltip only closes with another click on the item. 
```

mckaragoz committed:
Requested Changes and Landing Page GitHub Button Behavior

mckaragoz committed:
Merge branch 'TooltipReactWith' of https://github.com/mckaragoz/MudBlazor into TooltipReactWith

henon commented:
I expected that ClickToHide would hide the tooltip when directly clicking on the tooltip. Also, how can one toggle the tooltip by clicking? Can one use both `ClickToShow` and `ClickToHide` at the same time? It is impossible right now, right? Wouldn't a single option `ClickToToggle` be better than `ClickToShow` and `ClickToHide` ?

henon commented:
Actually a `bool ToggleOnClick` would be the better API then

mckaragoz commented:
> Actually a `bool ToggleOnClick` would be the better API then

Why not. But first i say why i add this, then you decide better. ClickToShow option directly means the toggle thing, ClickToHide is only for hide tooltip. There is two situation i don't know:

1. If there will be racing condition, is there a change to false open the tooltip again instead of closing?
2. Probably there will be more work about touch events in future, click pattern should not conflict with possible touch implementations.

Actually i like the bool parameter. If you say ok, i can change.

henon commented:
There can not be a race condition I think. OK, let's do the bool parameter

mckaragoz committed:
Change to ToggleOnClick bool Parameter

mckaragoz commented:
Changed to bool parameter and text of docs, could you look at these again?

henon commented:
The doc text is ok. The last thing I want to discuss are the side-effects of `ToggleOnClick`. It is impossible to have a hover tooltip that can be clicked away now?

henon commented:
I know it overrides the others. I just want to know if this is side-effect is absolutely necessary or if it would also work if it would not override the others. 

For instance, setting ShowOnFocus = true and ToggleOnClick = true, what if ToggleOnClick has no side-effect? Would it still work? Right now, you can not have both.

mckaragoz committed:
Change API again

mckaragoz commented:
@henon thanks, i like the new API now, its much clear and much fixable if there will be some bugs. I changed the names and behavior. Also changed GitHub and example "Run" button behavior to show tooltip properly.

henon commented:
@mckaragoz: So I guess the enum can be removed? And in the docs the `Note:` about the side effect can be removed.

I am not sure about the naming `Click`, `Hover` and `Focus`, it is not clear from the names what they do. `ShowOnClick`, `ShowOnHover` and `ShowOnFocus` is much more concise. You agree @mikes-gh ?

mikes-gh commented:
> @mckaragoz: So I guess the enum can be removed? And in the docs the `Note:` about the side effect can be removed.
> 
> I am not sure about the naming `Click`, `Hover` and `Focus`, it is not clear from the names what they do. `ShowOnClick`, `ShowOnHover` and `ShowOnFocus` is much more concise. You agree @mikes-gh ?

Yes you are right Click Hover Focus are actions not behaviours.

mikes-gh commented:
Also probably needs tests. Can I just be clear these parameters can all coexist?  ShowOnFocus in combination with ShowOnHover is good for accessibility for example with keyboard users without fine motor skills.

mckaragoz commented:
> Also probably needs tests. Can I just be clear these parameters can all coexist? ShowOnFocus in combination with ShowOnHover is good for accessibility for example with keyboard users without fine motor skills.

In current situation, tooltip already works on hover and on focus, but there is no click option to show/hide. The new one doesn't have breaking changes, it has hover and focus to true, and click to false by default, same behavior.

And yes, all three are possible at the same time, or possible all combinations.

This is important for me because for now there is no option to disable focus option without disabling all tooltip. With this everything is possible :)

mikes-gh commented:
>In current situation, tooltip already works on hover and on focus, but there is no click option to show/hide. The new one doesn't have breaking changes, it has hover and focus to true, and click to false by default, same behavior.

You've added logic thats not covered by the tests.

Also I would change the Example file name to be consistent with terminology.
TooltipReactExample -> TooltipActivationExample

mckaragoz commented:
> >In current situation, tooltip already works on hover and on focus, but there is no click option to show/hide. The new one doesn't have breaking changes, it has hover and focus to true, and click to false by default, same behavior.
> 
> You've added logic thats not covered by the tests.
> 
> Also I would change the Example file name to be consistent with terminology.
> TooltipReactExample -> TooltipActivationExample

Yeah I will handle all, but henon asks you for the new API naming, should we have "ShowOnHover" or simply "Hover"?

henon commented:
@mckaragoz all with ShowOn prefix please

mckaragoz committed:
Last Arrangement

mckaragoz commented:
@henon sorry, didn't see the mikes's message on upper, changed names and docs, removed enum, add a click test. And other hover and focus events tests work.

henon commented:
Many thanks!
